File: frontend/src/types/payroll.ts
export interface SalaryComponent {
  id: number;
  structure_id: number;
  name: string;
  type: "earning" | "deduction";
  calculation_type: "fixed" | "percentage_of_basic" | "percentage_of_gross";
  amount_or_percentage: number;
  is_mandatory: boolean;
  display_order: number;
}

export interface SalaryComponentCreate {
  name: string;
  type: "earning" | "deduction";
  calculation_type: "fixed" | "percentage_of_basic" | "percentage_of_gross";
  amount_or_percentage: number;
  is_mandatory: boolean;
  display_order: number;
}

export interface SalaryStructure {
  id: number;
  name: string;
  description: string | null;
  is_default: boolean;
  components: SalaryComponent[];
  created_at: string;
}

export interface EmployeeSalary {
  id: number;
  employee_id: number;
  structure_id: number;
  basic_pay: number;
  effective_from: string;
  effective_to: string | null;
}

export interface Advance {
  id: number;
  employee_id: number;
  amount: number;
  disbursed_date: string;
  repayment_months: number;
  monthly_deduction: number;
  remaining_balance: number;
  status: "active" | "repaid";
  notes: string | null;
  created_at: string;
}

export interface PayrollItem {
  id: number;
  run_id: number;
  employee_id: number;
  basic_pay: number;
  earnings_breakdown: Record<string, number>;
  deductions_breakdown: Record<string, number>;
  days_worked: number;
  overtime_hours: number;
  overtime_pay: number;
  gross_pay: number;
  total_deductions: number;
  advance_deduction: number;
  net_pay: number;
  status: string;
  employee_name?: string;
}

export interface PayrollRun {
  id: number;
  month: number;
  year: number;
  status: "draft" | "processing" | "completed" | "cancelled";
  run_by: number | null;
  started_at: string;
  completed_at: string | null;
  total_gross: number | null;
  total_net: number | null;
  total_deductions: number | null;
  employee_count: number | null;
}

export interface PayrollRunDetail extends PayrollRun {
  items: PayrollItem[];
}

File: backend/schemas/payroll.py
from pydantic import BaseModel, Field
from decimal import Decimal
from datetime import datetime, date
from typing import Optional, Literal


# ── Salary Structure ──────────────────────────────────────────────────────────

class SalaryComponentCreate(BaseModel):
    name:                 str = Field(..., max_length=100)
    type:                 Literal["earning", "deduction"]
    calculation_type:     Literal["fixed", "percentage_of_basic", "percentage_of_gross"]
    amount_or_percentage: Decimal = Field(..., ge=0)
    is_mandatory:         bool = True
    display_order:        int = 0


class SalaryComponentResponse(BaseModel):
    id:                   int
    structure_id:         int
    name:                 str
    type:                 str
    calculation_type:     str
    amount_or_percentage: Decimal
    is_mandatory:         bool
    display_order:        int

    model_config = {"from_attributes": True}


class SalaryStructureCreate(BaseModel):
    name:        str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = None
    is_default:  bool = False
    components:  list[SalaryComponentCreate] = []


class SalaryStructureUpdate(BaseModel):
    name:        Optional[str] = Field(None, max_length=255)
    description: Optional[str] = None
    is_default:  Optional[bool] = None
    components:  Optional[list[SalaryComponentCreate]] = None


class SalaryStructureResponse(BaseModel):
    id:          int
    name:        str
    description: Optional[str] = None
    is_default:  bool
    components:  list[SalaryComponentResponse] = []
    created_at:  datetime

    model_config = {"from_attributes": True}


# ── Employee Salary Assignment ────────────────────────────────────────────────

class EmployeeSalaryCreate(BaseModel):
    structure_id: int
    basic_pay:    Decimal = Field(..., gt=0)


class EmployeeSalaryResponse(BaseModel):
    id:             int
    employee_id:    int
    structure_id:   int
    basic_pay:      Decimal
    effective_from: datetime
    effective_to:   Optional[datetime] = None

    model_config = {"from_attributes": True}


# ── Advance ───────────────────────────────────────────────────────────────────

class AdvanceCreate(BaseModel):
    employee_id:      int
    amount:           Decimal = Field(..., gt=0)
    disbursed_date:   date
    repayment_months: int = Field(..., ge=1, le=60)
    notes:            Optional[str] = None


class AdvanceResponse(BaseModel):
    id:                int
    employee_id:       int
    amount:            Decimal
    disbursed_date:    date
    repayment_months:  int
    monthly_deduction: Decimal
    remaining_balance: Decimal
    status:            str
    notes:             Optional[str] = None
    created_at:        datetime

    model_config = {"from_attributes": True}


# ── Payroll Run ───────────────────────────────────────────────────────────────

class PayrollRunCreate(BaseModel):
    month: int = Field(..., ge=1, le=12)
    year:  int = Field(..., gt=2000)


class PayrollItemResponse(BaseModel):
    id:                   int
    run_id:               int
    employee_id:          int
    basic_pay:            Decimal
    earnings_breakdown:   dict
    deductions_breakdown: dict
    days_worked:          int
    overtime_hours:       Decimal
    overtime_pay:         Decimal
    gross_pay:            Decimal
    total_deductions:     Decimal
    advance_deduction:    Decimal
    net_pay:              Decimal
    status:               str
    employee_name:        Optional[str] = None

    model_config = {"from_attributes": True}


class PayrollRunResponse(BaseModel):
    id:               int
    month:            int
    year:             int
    status:           str
    run_by:           Optional[int] = None
    started_at:       datetime
    completed_at:     Optional[datetime] = None
    total_gross:      Optional[Decimal] = None
    total_net:        Optional[Decimal] = None
    total_deductions: Optional[Decimal] = None
    employee_count:   Optional[int] = None

    model_config = {"from_attributes": True}


class PayrollRunDetailResponse(PayrollRunResponse):
    items: list[PayrollItemResponse] = []
